opm/simulators/wells/WellBhpThpCalculator.cpp

```diff
@@ -288,32 +297,47 @@ calculateBhpFromThp(const WellState& well_state,
-        return well_.vfpProperties()->getInj()->bhp(controls.vfp_table_number, aqua, liquid, vapour, well_.getTHPConstraint(summaryState)) - dp;
-     }
-     else if (well_.isProducer()) {
-         const auto& controls = well.productionControls(summaryState);
-         const double vfp_ref_depth = well_.vfpProperties()->getProd()->getTable(controls.vfp_table_number).getDatumDepth();
-         const double dp = wellhelpers::computeHydrostaticCorrection(well_.refDepth(), vfp_ref_depth, rho, well_.gravity());
-         const auto& wfr =  well_.vfpProperties()->getExplicitWFR(controls.vfp_table_number, well_.indexOfWell());
-         const auto& gfr = well_.vfpProperties()->getExplicitGFR(controls.vfp_table_number, well_.indexOfWell());
-         const bool use_vfpexplicit = well_.useVfpExplicit();
-         return well_.vfpProperties()->getProd()->bhp(controls.vfp_table_number,
+        vfp_ref_depth = well_.vfpProperties()->getInj()->getTable(controls.vfp_table_number).getDatumDepth();
+        bhp_tab = well_.vfpProperties()->getInj()->bhp(
+               controls.vfp_table_number, aqua, liquid, vapour, thp_limit);
+    }
+    else if (well_.isProducer()) {
+        const auto& controls = well.productionControls(summaryState);
+        vfp_ref_depth = well_.vfpProperties()->getProd()->getTable(controls.vfp_table_number).getDatumDepth();
+        const auto& wfr =  well_.vfpProperties()->getExplicitWFR(controls.vfp_table_number, well_.indexOfWell());
+        const auto& gfr = well_.vfpProperties()->getExplicitGFR(controls.vfp_table_number, well_.indexOfWell());
+        const bool use_vfpexplicit = well_.useVfpExplicit();
+        bhp_tab = well_.vfpProperties()->getProd()->bhp(controls.vfp_table_number,
                                                       aqua, liquid, vapour,
-                                                      well_.getTHPConstraint(summaryState),
+                                                      thp_limit,
                                                       well_.getALQ(well_state),
-                                                      wfr, gfr, use_vfpexplicit) - dp;
-     }
-     else {
-         OPM_DEFLOG_THROW(std::logic_error, "Expected INJECTOR or PRODUCER for well " + well_.name(), deferred_logger);
-     }
+                                                      wfr, gfr, use_vfpexplicit);
+    }
+    else {
+        OPM_DEFLOG_THROW(std::logic_error, "Expected INJECTOR or PRODUCER for well " + well_.name(), deferred_logger);
+    }
+    double bhp_tab_double_value;
+    if constexpr (std::is_same_v<EvalWell, double>) {
+        bhp_tab_double_value = bhp_tab;
+    }
+    else {  // EvalWell and bhp_tab is of type DenseAd::Evaluation<double,...,...>
+        bhp_tab_double_value = bhp_tab.value();
+    }
+    const auto bhp_adjustment = getVfpBhpAdjustment(bhp_tab_double_value, thp_limit);
```

Comment:
It is better to use bhp_tab_double_value = getValue(bhp_tab) then you don't need the std::is_same check.

Reply:
Good idea. I assume the function `getValue()` is the one declared here: https://github.com/OPM/opm-material/blob/ee6c07fef239219d2470bc9be482ca894d2e60df/opm/material/common/MathToolbox.hpp#L330 ? If I try this the compiler says:

```
/home/hakon/test/opm11/opm-common/opm/material/common/MathToolbox.hpp:51:52: error: static assertion failed: This class expects floating point scalars! (specialization missing?)
   51 |     static_assert(std::is_floating_point<ScalarT>::value,
      |                                                    ^~~~~
/home/hakon/test/opm11/opm-common/opm/material/common/MathToolbox.hpp:51:52: note: ‘std::integral_constant<bool, false>::value’ evaluates to false
/home/hakon/test/opm11/opm-simulators/opm/simulators/wells/WellBhpThpCalculator.cpp: In instantiation of ‘EvalWell Opm::WellBhpThpCalculator::calculateBhpFromThp(const Opm::WellState&, const std::vector<EvalWell>&, const Opm::Well&, const Opm::SummaryState&, double, Opm::DeferredLogger&) const [with EvalWell = Opm::DenseAd::Evaluation<double, 3, 0>]’:
/home/hakon/test/opm11/opm-simulators/opm/simulators/wells/WellBhpThpCalculator.cpp:793:1:   required from here
/home/hakon/test/opm11/opm-simulators/opm/simulators/wells/WellBhpThpCalculator.cpp:326:66: error: cannot convert ‘Opm::MathToolbox<Opm::DenseAd::Evaluation<double, 3, 0> >::Scalar’ {aka ‘Opm::DenseAd::Evaluation<double, 3, 0>’} to ‘double’
  326 |     const auto bhp_adjustment = getVfpBhpAdjustment(Opm::getValue(bhp_tab), thp_limit);
      |                                                     ~~~~~~~~~~~~~^~~~~~~~~
      |                                                                  |
      |                                                                  Opm::MathToolbox<Opm::DenseAd::Evaluation<double, 3, 0> >::Scalar {aka Opm::DenseAd::Evaluation<double, 3, 0>}
/home/hakon/test/opm11/opm-simulators/opm/simulators/wells/WellBhpThpCalculator.cpp:332:63: note:   initializing argument 1 of ‘double Opm::WellBhpThpCalculator::getVfpBhpAdjustment(double, double) const’
  332 | double WellBhpThpCalculator::getVfpBhpAdjustment(const double bhp_tab, const double thp_limit) const
      |                                                  ~~~~~~~~~~~~~^~~~~~~
```
Do I need to give a template argument to `getValue()` or is it something else?

Reply:
getValue will convert from Evaluation to Scalar. Seems like the compiler complains about conversion from Scalar to double 

Try 

`double bhp_tab_double_value = getValue(bhp_tab);`

and than use bhp_tab_double_value in the functions.